service/history/timerGate.go

```diff
@@ -125,20 +130,28 @@ func (timerGate *LocalTimerGateImpl) FireAfter(now time.Time) bool {
 // Update update the timer gate, return true if update is a success
 // success means timer is idle or timer is set with a sooner time to fire
 func (timerGate *LocalTimerGateImpl) Update(nextTime time.Time) bool {
+	// NOTE: negative duration will make the timer fire immediately
 	now := time.Now()
 
-	if !timerGate.nextWakeupTime.After(now) || timerGate.nextWakeupTime.After(nextTime) {
-		// if timer will not fire or next wake time is after the "next"
-		// then we need to update the timer to fire
+	if timerGate.timer.Stop() {
+		// this means the timer, before stopped, is active
+
+		if timerGate.nextWakeupTime.Before(nextTime) {
+			timerGate.timer.Reset(timerGate.nextWakeupTime.Sub(now))
+			return false
+		}
+
 		timerGate.nextWakeupTime = nextTime
-		// reset timer to fire when the next message should be made 'visible'
 		timerGate.timer.Reset(nextTime.Sub(now))
-
 		// Notifies caller that next notification is reset to fire at passed in 'next' visibility time
 		return true
 	}
 
-	return false
+	// this means the timer, before stopped, is already fired / never active
```

Comment:
This logic is exactly same as the logic within the 'if timerGate.timer.Stop'.  Let's have it only outside of the if check.